### PR TITLE
Customize BMP manual entry UIs

### PR DESCRIPTION
**Pull request by lewfish**
This PR adds a configuration for each GWLF-E BMP to customize the manual entry UI for each BMP.

When reviewing this PR, ignore style issues including layout and word choices as they will be addressed in https://github.com/WikiWatershed/model-my-watershed/issues/1315.

Currently, this doesn't add a modification to the ScenarioModel, and uses a mock Mapshed data model. This will be addressed in https://github.com/WikiWatershed/model-my-watershed/issues/1316

#### Testing
* Check that BMP UI still works for TR55.
* For the BMP dropdown for GWLF-E models, check that back button and clicking the dropdown button works as expected (should clear things and bring back the thumbnails).
* For each BMP check that everything matches the spec at https://drive.google.com/open?id=0B8dzaPNE2DLTUV9EWC1tNXVVWDhUeHhPQzlIVlYyWVVJemxz
* Clicking Convert will not work if there are error messages.
* Clicking Convert when the input is valid will print an object to the console with the output.
screen shot 2016-05-18 at 11 16 30 am 2
screen shot 2016-05-18 at 11 16 52 am 2
screen shot 2016-05-18 at 11 16 59 am 2

Connects #1307

### CODE REVIEW

**Comment by lewfish** [comment]
@mmcfarland: For rural vegetated buffer strips, streambank fencing, and streambank stabilization, is the maximum allowable length n42 (total length in ag areas) or n42b (total length in watershed)?

File: src/mmw/js/src/modeling/gwlfeModificationConfig.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    QretentionDefault = 2.54;
+
+// This function converts a list of pairs to objects. We use it to
+// create objects where the keys are the values of variables, which is not
+// possible with object literal syntax. For instance,
+// x = {n23Name: 'y'} would create an object {'n23Name': 'y'} instead of
+// {'n23': 'y'} which is what we want. We could also do {n23: 'y'} but that
+// would defeat the point of using constants here which is to get
+// spell-checking from the interpreter.
+function toObject(pairs) {
```
Comment (caseypt) [inline src/mmw/js/src/modeling/gwlfeModificationConfig.js]: Does [_.fromPairs](https://lodash.com/docs#fromPairs) accomplish this?

Comment (lewfish) [inline src/mmw/js/src/modeling/gwlfeModificationConfig.js]: Yes! That's what I was looking for.

File: src/mmw/js/src/modeling/gwlfeModificationConfig.js [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    (received from the backend) which should be displayed to the user.
+    - userInputNames is a list of non-Mapshed variables used to reference values
+    the user enters into the UI. currently all the BMPs only reference a single
+    user input variable, so using a list is overkill, but it could be useful in
+    the future.
+    - validate is a function which validates the user input using values from
+    the data model.
+    - computeOutput is a function which computes the output of the UI for a BMP,
+    the values for a set of Mapshed variables.
+*/
```
Comment (caseypt) [inline src/mmw/js/src/modeling/gwlfeModificationConfig.js]: Nice comment. I think on future projects we should consider using something like [JSDoc](http://usejsdoc.org/about-getting-started.html)

**Comment by caseypt** [comment]
For validation, consider using [HTML5 form validation](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Forms/Data_form_validation).

**Comment by caseypt** [comment]
Good documentation on the complicated stuff.

**Comment by caseypt** [comment]
Verified that
- TR-55 modification control still works
- You can't enter values if the input is in valid
- Variables output names for each modification line up with the document.

**Comment by caseypt** [comment]
If you set a value for one modification type, then open it again, should we display the current value? There is only one modification be "added" per type, not multiple, right?

**Comment by caseypt** [comment]
This is more for https://github.com/WikiWatershed/model-my-watershed/issues/1315, but if the above assumption is correct, we may want to show the current value for each activate modification on the modification control instead of the list of modifications.

**Comment by caseypt** [comment]
Talked over those last two comments with @lewfish. They will be addressed in #1316

**Comment by caseypt** [comment]
+1. Nice work.

**Comment by lewfish** [comment]
> Talked over those last two comments with @lewfish. They will be addressed in #1316

After talking to @mmcfarland, we decided to just leave things as they are, which means that clicking a BMP that has already been added will let the user reset the value, but will not show them the pre-existing value.

**Comment by lewfish** [comment]
> For validation, consider using HTML5 form validation.

I think it'll be too big of a change for this PR, but it's interesting to know about for next time. It looks like the validation is based on regular expressions which isn't the best fit for this PR.

**Comment by mmcfarland** [comment]
@lewfish regarding max stream length for the three km based rural BMPs, I believe it is n42 (streams in Ag), but I've got an email out to confirm.

**Pull request merged**